ui/src/tempVars/components/KeysTemplateBuilder.tsx

```diff
@@ -0,0 +1,241 @@
+                onChoose={this.handleChooseMeasurementDropdown}
+                selected={selectedMeasurement}
+                buttonSize=""
+              />
+            </DropdownPlaceholder>
+          </div>
+        </div>
+        <TemplateMetaQueryPreview items={keys} loadingState={keysStatus} />
+      </div>
+    )
+  }
+
+  private async loadDatabases() {
+    const {source} = this.props
+
+    this.setState({databasesStatus: RemoteDataState.Loading})
+
+    try {
+      const {data} = await showDatabases(source.links.proxy)
+      const {databases} = parseShowDatabases(data)
+      const {selectedDatabase} = this.state
+
+      this.setState({
+        databases,
+        databasesStatus: RemoteDataState.Done,
+      })
+
+      if (!selectedDatabase) {
+        this.handleChooseDatabase(_.get(databases, 0, ''))
+      }
+    } catch (error) {
+      this.setState({databasesStatus: RemoteDataState.Error})
+      console.error(error)
+    }
+  }
+
+  private async loadMeasurements() {
+    const {source} = this.props
+    const {selectedDatabase, selectedMeasurement} = this.state
+
+    this.setState({measurementsStatus: RemoteDataState.Loading})
+
+    try {
+      const {data} = await showMeasurements(
+        source.links.proxy,
+        selectedDatabase
+      )
+      const {measurementSets} = parseShowMeasurements(data)
+      const measurements = _.get(measurementSets, '0.measurements', [])
+
+      this.setState({
+        measurements,
+        measurementsStatus: RemoteDataState.Done,
+      })
+
+      if (!selectedMeasurement) {
+        this.handleChooseMeasurement(_.get(measurements, 0, ''))
+      }
+    } catch (error) {
+      this.setState({measurementsStatus: RemoteDataState.Error})
+      console.error(error)
+    }
+  }
+
+  private async loadKeys() {
+    const {
+      template,
+      onUpdateTemplate,
+      templateValueType,
+      fetchKeys,
+      source,
+    } = this.props
+
+    const {selectedDatabase, selectedMeasurement} = this.state
+
+    this.setState({keysStatus: RemoteDataState.Loading})
+
+    try {
+      const keys = await fetchKeys(
+        source,
+        selectedDatabase,
+        selectedMeasurement
+      )
+
+      this.setState({
+        keys,
+        keysStatus: RemoteDataState.Done,
+      })
+
+      const nextValues = keys.map(value => {
+        return {
+          type: templateValueType,
+          value,
+          selected: false,
+        }
+      })
+
+      if (nextValues[0]) {
+        nextValues[0].selected = true
+      }
+
+      onUpdateTemplate({...template, values: nextValues})
+    } catch (error) {
+      this.setState({keysStatus: RemoteDataState.Error})
+      console.error(error)
+    }
+  }
+
+  private handleChooseDatabaseDropdown = ({text}) => {
+    this.handleChooseDatabase(text)
+  }
+
+  private handleChooseDatabase = (db): void => {
+    this.setState({selectedDatabase: db, selectedMeasurement: ''}, () =>
```

Comment:
just thinking out loud, but might setState return a promise, in which case this could use `async/await`?